### ISSUE TITLE
How to query a view instead of a table in sqlite3?

### ISSUE THREAD

**Issue by Deilan** [typeorm/typeorm#1024]
Let's say I have sqlite3 and an entity `Person`. I want to create a view `PersonView` with some calculable data along with original data in `Person`. How to query a view instead of a table with typeorm in sqlite3?

**Comment by pleerock** [comment]
No, typeorm does not provide any additional functionality to work with views. You can create a proposal how do you see this functionality in typeorm and we can discuss it.

**Comment by xipheCom** [comment]
A view with typorm would be a dream. 

My first thoughts:

In ERM a view is just a query. A view could be a query in typeorm too.

``` typescript
@View('SpecialUserView') //<<--optional view name
AdminUserView() => { //<<--function name = view name for database if not set in @View
  return getRepository(User)
    .createQueryBuilder("user")
    .where("user.id = :id", { id: 1 })
    .getOne();
}
```

We should be able to make query from a view. 
``` typescript
getView(AdminUserView) //instead of getRepository
    .createQueryBuilder("user")
    .where("user.active = :active", { active: true })
    .getOne();
```

What do you think @pleerock ?

**Comment by xipheCom** [comment]
Second: A view in an orm is most of time a subset of an entity. I think this is not the correct syntax and I do not know if we could do it with decorators this way, but you get the point:

```typescript
@Entity()
export class Product {
    PrimaryColumn()
    id: number;

    Column()
    productNumber: string;

    Column()
    title: string;

    Column()
    description: string;

    Column()
    someAdminStuff: string;

    Column()
    other2: string;

    //....
}

//mobile view
@View()  //or call it subset or so....
export class ProductView {
    id: Product.id;

    title: Product.title;

}
```

**Comment by pleerock** [comment]
@xipheCom thanks for the feedback! First approach looks better - its flexible, cover all cases, easy to implement and does not introduce overhead. 

```ts
getView(AdminUserView) //instead of getRepository
    .createQueryBuilder("user")
    .where("user.active = :active", { active: true })
    .getOne();
``

what other methods `getView` can contain? Can we do something else rather then querying with views? If no, then `createQueryBuilder` is enough, e.g. `createQueryBuilder(UserView, "user")`.

Looks like easy feature to implement, are you able to contribute?

**Comment by xipheCom** [comment]
@pleerock You are right, there are no benefits from a "getView" if we can use it as a parameter in createQueryBuilder. So we can make (even cascading or "nested") views and everyone should be happy.

Unfortunatly this is very new for me. Developing my first project with typorm just now. I have created my first decorator just a few days ago. So I can not contribute with coding in next months. But if you make the code I will write the documentation as a time saver for you. Would make a PR if code is ready. Topics would be:

* Views 
  Introduction - Views are just queries
* Create a View 
  with @View Decorator, function name is view name in database (note: naming strategy), View naming with @View('special_name')
* Use a view
  example
*  Cascading / Nested views
   short sentence about nested views in general, example of using

**Comment by pleerock** [comment]
Okay, then let's see if someone else is interesting in implementing this. Its easy to implement feature. Thanks for the design proposal.

**Comment by chcas** [comment]
In my case it would be very interesting to build the result of a search without having to use QueryBuilder

**Comment by Mijago** [comment]
Any progress here? 
I'd love to see this feature, so _maybe_ i can find some time to implement it or at least do some tests.

**Comment by mdostal** [comment]
Yeah, I'm able to attach to synonyms and views in MSSQL.  Currently I'm attaching primary column to all columns, will see if I can limit down to true uniqueness.   But this seems to be working for me.

**Comment by romansflipsidegroup** [comment]
I have created normal @Entity, like for tables, but using it on MariaDB 10.3 view. No problems to make selects. Did someone tried it on PostgreSQL?

**Comment by romansflipsidegroup** [comment]
I have created normal @Entity, like for tables, but using it on MariaDB 10.3 view. No problems to make selects. Did someone tried it on PostgreSQL?

**Comment by romanszedzielorz** [comment]
I have created normal TypeORM @Entity, like for tables, but using it on MariaDB 10.3 view. No problems to make selects. Did someone tried it on PostgreSQL?

**Comment by Phrohdoh** [comment]
In a few projects of mine using PostgreSQL I have tables that only have `JSONB` columns which of course do not have primary keys.

It would be nice if typeorm supported this out-of-the-box but while it doesn't I can of course create a "throwaway" PK column.

**Comment by Maxpain177** [comment]
+1

**Comment by P0oOOOo0YA** [comment]
Views are an unavoidable feature of any serious project. After two personal project now i'm going to use `typeorm` in production and certainly need to work with views. @pleerock  What are the latest news on this feature? Is it implemented or is there any guideline or recommendation?

**Comment by P0oOOOo0YA** [comment]
@romanszedzielorz could you please elaborate more on what you have exactly done? How did you create the view in the first place?

**Comment by P0oOOOo0YA** [comment]
@romanszedzielorz thanks i had forgotten migrations totally :) .

**Comment by AlexMesser** [comment]
View entity implemented and will be released in `0.2.17`. Here is the [docs](https://typeorm.io/#/view-entities)

**Comment by smtaha512** [comment]
Hi. I can not find the `0.2.17` version of typeorm mentioned above on npm.

**Comment by Phrohdoh** [comment]
> and _will be_ released in `0.2.17`

@smtaha512, It hasn't been published yet.

**Comment by smtaha512** [comment]
Any expected timeline?

**Comment by pleerock** [comment]
Think I'll publish it end of this week

**Comment by tzarger** [comment]
Is this still planned to be pushed at the end of this week? @pleerock

**Comment by pleerock** [comment]
awaiting few PRs to be completely reviewed and merged

**Comment by pleerock** [comment]
`0.2.17` was released

**Comment by kordeviant** [comment]
quick question: for using already created views we should just delete the creation step in migrations and continue?

**Comment by Firaenix** [comment]
I'm getting errors in 0.3.0-alpha that say ViewEntity doesnt exist in TypeORM anymore. Anyone else getting this?

**Comment by gsmental** [comment]
const data = await PostgresDataSource.createQueryBuilder()
    .select("*")
    .from("invent.vmasterproduct", "p")
    .getRawMany();

**Comment by shobhaeventhq** [comment]
Give me suggestion over use of view with typeorm , postgresql and also work with leftjoin when i have own querybuilder and and also i have persistance services and also can i will be able to use view entity with typeorm entity?
